engine/resources/consul_kv.go

```diff
@@ -0,0 +1,186 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util"
+)
+
+func init() {
+	engine.RegisterResource("consul:kv", func() engine.Res { return &ConsulKVRes{} })
+}
+
+// ConsulKVRes is a resource that writes a value into a consul datastore.
+type ConsulKVRes struct {
+	traits.Base
+	init *engine.Init
+
+	// Key is the path of the value.
+	Key string `lang:"key" yaml:"key"`
+
+	// Key is the value that will be written into consul.
+	Value string `lang:"value" yaml:"value"` // the consul key value
+
+	// Address is the address of the consul server.
+	Address string `lang:"address" yaml:"address"` // the consul server
+
+	// Token is the token used to connect to the consul server.
+	Token string `lang:"token" yaml:"token"` // the consul token
+
+	client *api.Client
+	config *api.Config // needed to close the idle connections
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *ConsulKVRes) Default() engine.Res {
+	return &ConsulKVRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *ConsulKVRes) Validate() error {
+	if obj.Key == "" {
+		return errors.New("key is mandatory")
+	}
+	return nil
+}
+
+// Init runs some startup code for this resource.
+func (obj *ConsulKVRes) Init(init *engine.Init) error {
+	obj.init = init // save for later
+
+	obj.config = api.DefaultConfig()
+	if obj.Address != "" {
+		obj.config.Address = obj.Address
+	}
+	if obj.Token != "" {
+		obj.config.Token = obj.Token
+	}
+
+	var err error
+	obj.client, err = api.NewClient(obj.config)
+	return err
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *ConsulKVRes) Close() error {
+	obj.config.Transport.CloseIdleConnections()
+	return nil
+}
+
+// Watch is the listener and main loop for this resource and it outputs events.
+func (obj *ConsulKVRes) Watch() error {
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+
+	c := make(chan error)
+	kv := obj.client.KV()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		opts := &api.QueryOptions{RequireConsistent: true}
+		ctx, cancel := util.ContextWithCloser(context.Background(), obj.init.Done)
+		defer cancel()
+		opts = opts.WithContext(ctx)
+
+	POLL:
+		_, meta, err := kv.Get(obj.Key, opts)
+		select {
+		case c <- err:
+			if err == nil {
+				if opts.WaitIndex == 0 {
+					// WaitIndex = 0, which means that it is the first time we
+					// run the query. as we are about to change the WaitIndex to
+					// make a blocking query, we can consider the watch started.
+					obj.init.Running()
+				}
+				opts.WaitIndex = meta.LastIndex
+				goto POLL
+			}
+		case <-obj.init.Done:
+		}
+	}()
+
+	for {
+		select {
+		case err := <-c:
+			if err == nil {
+				obj.init.Event()
+			} else {
```

Comment:
inline this

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow